### PR TITLE
Some improvements

### PR DESCRIPTION
**Pull request by rcatolino**
- The first commit just moves everything that needs linking against libpam.so into libpam.rs for consistency.
- The second commit makes the test module use some of the libpam features.
- The third commit replaces the ```get_item``` function for a safer ```get_cstr_item``` since all the items we use are actually ```CStr``` except for ```PAM_CONV```.  This simplify the rest of ```PamLibExt``` somewhat.
- The last commit fixes a bug where the CString allocated for the prompts in ```get_authtoken``` and ```get_user``` would get dropped too early and resulted in dangling pointers being sent to the user of the module.

### CODE REVIEW

File: src/libpam.rs [hunk 1]
```diff
@@ -1,69 +1,190 @@
...
-    let r = unsafe { PamError::new(pam_get_item(pamh, item_type as c_int, &mut raw_item)) };
-    if raw_item.is_null() {
-        r.to_result(None)
-    } else {
-        r.to_result(Some(raw_item))
-    }
+/// Extension trait over `Pam`, usually provided by the `libpam` shared library.
+pub trait PamLibExt: private::Sealed {
+    #[doc(hidden)] // End users should call the item specific methods
+    fn get_cstr_item(&self, item_type: PamItemType) -> PamResult<Option<&CStr>>;
```
Comment (lucab) [inline src/libpam.rs]: What about moving this function out of this and directly into a `impl Pam {}` block, in this same file? That way it won't be part of the public interface at all.

File: src/libpam.rs [hunk 2]
```diff
@@ -1,69 +1,190 @@
...
+    fn get_cached_user(&self) -> PamResult<Option<&CStr>> {
+        self.get_cstr_item(PamItemType::USER)
+    }
+
+    fn get_cached_authtok(&self) -> PamResult<Option<&CStr>> {
+        self.get_cstr_item(PamItemType::AUTHTOK)
     }
+
+    fn get_authtok(&self, prompt: Option<&str>) -> PamResult<Option<&CStr>> {
+        let cprompt = prompt.map(|p| CString::new(p).expect(ERR_CSTR_NULL));
```
Comment (lucab) [inline src/libpam.rs]: While at it, should this `expect` be massaged into proper error bubbling?

**Comment by rcatolino** [comment]
I've moved ```get_item_cstr``` into a ```impl Pam```, it makes a lot more sense indeed.

Allowing the service module to handle error caused by prompts with nul bytes could be useful if the prompt doesn't come from the service module but from an external source. I decided to return a ```PamError::SERVICE_ERR``` cause it's the only one that makes it clear the error comes from the service module and not the application. Ideally it would be a different type since it's not really a pam error but I think adding one just for this would be overkill.

**Review by lucab — APPROVED** [review]
LGTM

**Pull request merged**